components/raftstore/src/store/peer_storage.rs

```diff
@@ -767,12 +821,188 @@ where
         Ok(())
     }
 
+    pub fn clean_async_fetch_res(&mut self, low: u64) {
+        self.async_fetch_results.borrow_mut().remove(&low);
+    }
+
+    pub fn set_async_fetch_res(&mut self, low: u64, res: Option<RaftlogFetchResult>) {
+        let in_fetching = match self.async_fetch_results.borrow().get(&low) {
+            Some(RaftlogFetchState::Fetched(_)) => false,
+            Some(RaftlogFetchState::Fetching) => true,
+            None => false,
+        };
+
+        match res {
+            Some(res) => {
+                if let Some(RaftlogFetchState::Fetched(prev)) = self
+                    .async_fetch_results
+                    .borrow_mut()
+                    .insert(low, RaftlogFetchState::Fetched(res))
+                {
+                    warn!(
+                        "unconsumed async fetch res";
+                        "region_id" => self.region.get_id(),
+                        "peer_id" => self.peer_id,
+                        "res" => ?prev,
+                        "low" => low,
+                    );
+                }
+            }
+            None => {
+                if !in_fetching {
+                    let prev = self.async_fetch_results.borrow_mut().remove(&low);
+                    if prev.is_some() {
+                        self.raftlog_fetch_stats.fetch_invalid.update(|m| m + 1);
+                    }
+                }
+            }
+        }
+    }
+
+    fn async_fetch(
+        &self,
+        region_id: u64,
+        low: u64,
+        mut high: u64,
+        max_size: u64,
+        context: GetEntriesContext,
+        buf: &mut Vec<Entry>,
+    ) -> raft::Result<usize> {
+        if let Some(RaftlogFetchState::Fetching) = self.async_fetch_results.borrow().get(&low) {
+            // already an async fetch in flight
+            return Err(raft::Error::Store(
+                raft::StorageError::LogTemporarilyUnavailable,
+            ));
+        }
```

Comment:
If there is more than one peer requesting log entries start from `low` then only one peer will get the logs by async fetching and other peers may have to wait until raft heartbeat to trigger getting logs for another peer.

Reply:
Yes, that's a defect though I think it's rare. Using `low+high` as the map index could be a solution, but when high is changed the result can't be reused which I think it's a more common case.

At the very previous implementation, peer id is used as the index. But to make the context transparent to the application, we just drop it.

Reply:
> At the very previous implementation, peer id is used as the index. But to make the context transparent to the application, we just drop it.

Sorry, I don't get it. The peer id index is used for `async_fetch_results` and `async_fetch_results` is the application's internal state.

Reply:
the `to_peer_id` is not the application's internal state. We need to get the peer id to which the entries are sent from raft.

Reply:
How about implementing the `Hash` trait for `GetEntriesContext` (namely hash the `to_peer_id`) and using the hash value (an `u64`) for indexing.

Reply:
Can't. In other context, there is no peer id, e.g. `GenReady`